package-lock.json

```diff
@@ -0,0 +1,8485 @@
+{
+  "name": "solc",
+  "version": "0.8.10",
+  "lockfileVersion": 2,
```

Comment:
This is a library, not an application so I don't think having a `package.lock` is a good idea. A library is meant to work with a range of versions. For example if we say something like `xyz: ^4.0.0` in `package.json`, we should work with the latest version of `xyz` matching `^4.0.0` and not only `4.0.3` because that happened to be the current version last time we updated the lock file.

Reply:
Yeap, makes sense and something I jumped the gun for. I agree it should only be committed for top level projects.

Reply:
package-lock has now been removed.

Reply:
I missed this conversation previously, and I strongly disagree with the conclusion.  package.lock files ensure that you get reliable builds every time and it doesn't change outside of your control just because some dependency updated.  Lockfiles also provide cryptographic guarantees that the dependencies being used are the ones you used originally, which can protect against supply chain attacks.

If you write a library that is shown to work with version X of dependency Y, it can be incredibly frustrating for your users to have to troubleshoot why things are broken because they ended up using version Z of dependency Y.

Reply:
Are we happy to include the package-lock.json in this PR ?

Reply:
I would agree if the lock file was actually carefully reviewed every time it changes. I don't think that's the case in general. It's too big, too verbose and changes too often. It's very easy to sneak through a malicious change among the noise and it will override what `package.json` says. I think that relying on `package.json` as a single source of truth is safer in our case.

Also, I don't think that locking package versions in development actually helps when your users are not locked to the same versions. `package-lock.json` is not published in npm so the only place where versions would actually be locked is this repo. This will just make us miss problems that affects users.

In any case, I think this is a longer discussion and not at all related to the TypeScript migration. `package-lock.json` can be used in JS as well and it's already excluded from the repo so if it's an issue, we already have it. If you feel strongly about it, maybe you could open a separate issue on the topic? I personally think we should not version the file but this would give the discussion more visiblity so that others can state their opinions too.

Reply:
> it's already excluded from the repo so if it's an issue, we already have it.

This PR introduces `package-lock.json` to the `.gitignore` file, which suggests that it *is not* already being ignored by this repository.  I agree this is worthy of a longer discussion.

Reply:
I meant that it the file is already not being tracked in the repository and in that sense it's being ignored. The only reason I would not add it to `.gitignore` because it makes it harder to actually ignore it in practice - by deleting it whenever it appears.